staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go

```diff
@@ -0,0 +1,570 @@
+		curQueueLength >= qs.config.QueueLengthLimit {
+		return false
+	}
+
+	qs.enqueue(request)
+	return true
+}
+
+func (qs *queueSet) roundRobinQueue() int {
+	qs.robinIndex = (qs.robinIndex + 1) % len(qs.queues)
+	return qs.robinIndex
+}
+
+// selectQueue selects the minimum virtualFinish time from the set of queues
+// the starting queue is selected via roundrobin
+func (qs *queueSet) selectQueue() *fq.Queue {
+	minVirtualFinish := math.Inf(1)
+	var minQueue *fq.Queue
+	var minIndex int
+	for range qs.queues {
+		index := qs.roundRobinQueue()
+		queue := qs.queues[index]
+		if len(queue.Requests) != 0 {
+			currentVirtualFinish := queue.GetVirtualFinish(0, qs.estimatedServiceTime)
+			if currentVirtualFinish < minVirtualFinish {
+				minVirtualFinish = currentVirtualFinish
+				minQueue = queue
+				minIndex = index
+			}
+		}
+	}
+	qs.robinIndex = minIndex
+	return minQueue
+}
+
+// dequeue dequeues a request from the queueSet
+func (qs *queueSet) dequeue() (*fq.Request, bool) {
+	queue := qs.selectQueue()
+	if queue == nil {
+		return nil, false
+	}
+	request, ok := queue.Dequeue()
+	if !ok {
+		return nil, false
+	}
+	// When a request is dequeued for service -> qs.VirtualStart += G
+	queue.VirtualStart += qs.estimatedServiceTime
+	request.StartTime = qs.clock.Now()
+	// request dequeued, service has started
+	queue.RequestsExecuting++
+	metrics.UpdateFlowControlRequestsExecuting(qs.config.Name, queue.RequestsExecuting)
+	qs.numRequestsEnqueued--
+	return request, ok
+}
+
+// dequeueWithChannelAsMuchAsPossible runs a loop, as long as there
+// are non-empty queues and the number currently executing is less than the
+// assured concurrency value.  The body of the loop uses the fair queuing
+// technique to pick a queue, dequeue the request at the head of that
+// queue, increment the count of the number executing, and send true
+// to the request's channel.
+func (qs *queueSet) dequeueWithChannelAsMuchAsPossible() {
+	for qs.numRequestsEnqueued != 0 && qs.getRequestsExecutingLocked() < qs.config.ConcurrencyLimit {
+		_, ok := qs.dequeueWithChannel()
+		if !ok {
+			break
+		}
+	}
+}
+
+// dequeueWithChannel is convenience method for dequeueing requests that
+// require a message to be sent through the requests channel
+// this is a required pattern for the QueueSet the queueSet supports
+func (qs *queueSet) dequeueWithChannel() (*fq.Request, bool) {
+	req, ok := qs.dequeue()
+	if !ok {
+		return nil, false
+	}
+	qs.counter.Add(1)
+	req.DequeueChannel <- true
+	close(req.DequeueChannel)
+	return req, ok
+}
+
+// removeQueueAndUpdateIndexes uses reslicing to remove an index from a slice
+// and then updates the 'Index' field of the queues to be correct
+func removeQueueAndUpdateIndexes(queues []*fq.Queue, index int) []*fq.Queue {
+	removedQueues := append(queues[:index], queues[index+1:]...)
+	for i := index; i < len(removedQueues); i++ {
+		removedQueues[i].Index--
+	}
+	return removedQueues
```

Comment:
Are you returning the queues that have been removed, or the queues that are going to be kept?

Reply:
The kept queus.  I've updated this to read more appropriately- s/removedQueues/keptQueues